packages/react-dom/src/__tests__/ReactTreeTraversal-test.js

```diff
@@ -0,0 +1,304 @@
+      );
+
+      expect(mockFn.mock.calls).toEqual(expectedCalls);
+    });
+
+    it('should traverse two phase at shallowest node', () => {
+      var node = document.getElementById('P');
+      node.dispatchEvent(
+        new MouseEvent('click', {bubbles: true, cancelable: true}),
+      );
+
+      var expectedCalls = [
+        ['P', 'captured', 'click', ARG],
+        ['P', 'bubbled', 'click', ARG],
+      ];
+      expect(mockFn.mock.calls).toEqual(expectedCalls);
+    });
+  });
+
+  describe('Enter leave traversal', () => {
+    it('should not traverse when enter/leaving outside DOM', () => {
+      var leaveNode = document.createElement('div');
+      var enterNode = document.createElement('div');
+
+      document.body.appendChild(leaveNode);
+      document.body.appendChild(enterNode);
+
+      leaveNode.dispatchEvent(
+        new MouseEvent('mouseout', {
+          bubbles: true,
+          cancelable: true,
+          relatedTarget: enterNode,
+        }),
+      );
+
+      expect(mockFn).not.toHaveBeenCalled();
+    });
+
+    it('should not traverse if enter/leave the same node', () => {
+      var leaveNode = document.getElementById('P_P1_C1__DIV_1');
+      var enterNode = document.getElementById('P_P1_C1__DIV_1');
+
+      leaveNode.dispatchEvent(
+        new MouseEvent('mouseout', {
+          bubbles: true,
+          cancelable: true,
+          relatedTarget: enterNode,
+        }),
+      );
+
+      expect(mockFn).not.toHaveBeenCalled();
+    });
+
+    it('should traverse enter/leave to sibling - avoids parent', () => {
+      var leaveNode = document.getElementById('P_P1_C1__DIV_1');
+      var enterNode = document.getElementById('P_P1_C1__DIV_2');
+
+      var expectedCalls = [
+        ['P_P1_C1__DIV_1', 'mouseleave', ARG],
+        // enter/leave shouldn't fire anything on the parent
+        ['P_P1_C1__DIV_2', 'mouseenter', ARG2],
+      ];
+
+      leaveNode.dispatchEvent(
+        new MouseEvent('mouseout', {
+          bubbles: true,
+          cancelable: true,
+          relatedTarget: enterNode,
+        }),
+      );
+
+      expect(mockFn.mock.calls).toEqual(expectedCalls);
+    });
+
+    it('should traverse enter/leave to parent - avoids parent', () => {
+      var leaveNode = document.getElementById('P_P1_C1__DIV_1');
+      var enterNode = document.getElementById('P_P1_C1__DIV');
+
+      var expectedCalls = [['P_P1_C1__DIV_1', 'mouseleave', ARG]];
+
+      leaveNode.dispatchEvent(
+        new MouseEvent('mouseout', {
+          bubbles: true,
+          cancelable: true,
+          relatedTarget: enterNode,
+        }),
+      );
+
+      expect(mockFn.mock.calls).toEqual(expectedCalls);
+    });
+
+    it('should enter from the window', () => {
+      var leaveNode = document.createElement('div');
+      var enterNode = document.getElementById('P_P1_C1__DIV');
+
+      document.body.appendChild(leaveNode); // From the window or outside of the React sandbox.
```

Comment:
If we change nodes to always be added, tests like this will have to *remove* one instead.

Reply:
Could you elaborate? I don't follow.

Reply:
Never mind, I misread the tests.